functions/functions.go

```diff
@@ -0,0 +1,187 @@
+// Copyright ©2015 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Package functions provides objective functions for testing optimization
+// algorithms.
+//
+// We encourage outside contributions of additional test functions that exhibit
+// properties not already covered in the testing suite or with significance due
+// to use in benchmark cases.
+package functions
+
+import (
+	"math"
+	"testing"
+
+	"github.com/gonum/diff/fd"
+	"github.com/gonum/floats"
+)
+
+// Function represents an objective function.
+type Function interface {
```

Comment:
These interfaces can all be private until we decide on the proper API to export.